### PR TITLE
dyn-config: read the config file line by line instead of allocating a fixed 32K buffer

### PR DESCRIPTION
**Pull request by lxbsz**
Fixes: #120

### CODE REVIEW

**Comment by lxbsz** [comment]
@pkalever 

Please review. thanks，

File: utils/dyn-config.c [hunk 1]
```diff
@@ -195,60 +197,84 @@ glusterBlockConfFreeStrKeys(gbConfig *cfg)
...
+  while ((m = getline(&line, &n, fp)) != -1) {
+    if (gluserBlockIsBlankOrCommentLine(line, m)) {
+      continue;
+    }
+
+    *len += m;
+    if (*len > buf_len) {
+      buf_len += GB_BUF_LEN;
+      if (GB_REALLOC_N(buf, buf_len) < 0) {
+	      free(line);
```
[the remark this answers is not in the document]
Comment (lxbsz) [reply utils/dyn-config.c]: Yeah, fixed.

File: utils/dyn-config.c [hunk 2]
```diff
@@ -195,60 +197,84 @@ glusterBlockConfFreeStrKeys(gbConfig *cfg)
...
-        } while (0);
+  *len = 0;
+  p = buf;
+  while ((m = getline(&line, &n, fp)) != -1) {
+    if (gluserBlockIsBlankOrCommentLine(line, m)) {
+      continue;
+    }
+
+    *len += m;
+    if (*len > buf_len) {
```
[the remark this answers is not in the document]
Comment (lxbsz) [reply utils/dyn-config.c]: Yeah, it should be.

File: utils/dyn-config.c [hunk 3]
```diff
@@ -195,60 +197,84 @@ glusterBlockConfFreeStrKeys(gbConfig *cfg)
...
-#define GB_SKIP_COMMENT_LINE(x, y) \
-        do { \
-          while ((x) < (y) && !__EOL(*x)) \
-            { (x)++; } \
-          (x)++; \
-        } while (0);
+  fp = fopen(cfg->configPath, "r");
+  if (fp == NULL) {
+    LOG("mgmt", GB_LOG_ERROR,
+        "Failed to open file '%s', %m\n", cfg->configPath);
```
[the remark this answers is not in the document]
Comment (lxbsz) [reply utils/dyn-config.c]: will fix.

File: utils/dyn-config.c [hunk 4]
```diff
@@ -405,40 +428,24 @@ glusterBlockParseOptions(gbConfig *cfg, char *buf, int len, bool reloading)
...
-  ret = 0;
-free_buf:
-  GB_FREE(buf);
-  return ret;
+  free(buf);
+  return 0;
 }
 
-#define BUF_LEN 1024
+# define BUF_LEN 1024
```
[the remark this answers is not in the document]
Comment (lxbsz) [reply utils/dyn-config.c]: Will reuse it.

**Comment by lxbsz** [comment]
@pkalever 
Updated it. Please review. Thanks.

File: utils/dyn-config.c [hunk 5]
```diff
@@ -196,60 +198,87 @@ glusterBlockConfFreeStrKeys(gbConfig *cfg)
...
+    GB_STRCPY(p, line, m + 1);
+    p += m;
+    *len += m;
+  }
+
+  *len += 1;
+  buf[*len] = '\0';
+
+  fclose(fp);
+  free(line);
```
Comment (pkalever) [inline utils/dyn-config.c]: USE GB_FREE() instead ?

File: utils/dyn-config.c [hunk 6]
```diff
@@ -196,60 +198,87 @@ glusterBlockConfFreeStrKeys(gbConfig *cfg)
...
+  p = buf;
+  while ((m = getline(&line, &n, fp)) != -1) {
+    if (gluserBlockIsBlankOrCommentLine(line, m)) {
+      continue;
+    }
+
+    if (*len + m >= buf_len) {
+      buf_len += GB_BUF_LEN;
+      if (GB_REALLOC_N(buf, buf_len) < 0) {
+	      free(line);
```
Comment (pkalever) [inline utils/dyn-config.c]: USE GB_FREE() instead ?

File: utils/dyn-config.c [hunk 7]
```diff
@@ -406,46 +432,29 @@ glusterBlockParseOptions(gbConfig *cfg, char *buf, int len, bool reloading)
...
 
-  buf[len] = '\0';
-
   glusterBlockParseOptions(cfg, buf, len, reloading);
 
-  ret = 0;
-free_buf:
-  GB_FREE(buf);
-  return ret;
+  free(buf);
```
Comment (pkalever) [inline utils/dyn-config.c]: USE GB_FREE() instead ?

File: utils/dyn-config.c [hunk 8]
```diff
@@ -195,60 +197,84 @@ glusterBlockConfFreeStrKeys(gbConfig *cfg)
...
+    if (*len > buf_len) {
+      buf_len += GB_BUF_LEN;
+      if (GB_REALLOC_N(buf, buf_len) < 0) {
+	      free(line);
+	      return NULL;
+      }
+      p = buf + *len - m;
+    }
+    GB_STRCPY(p, line, m + 1);
+    p += m;
```
[the remark this answers is not in the document]
Comment (pkalever) [reply utils/dyn-config.c]: OK

**Comment by pkalever** [comment]
Rest looks good to me. Thanks @lxbsz

**Comment by lxbsz** [comment]
@pkalever 
Done, I think I have addressed them all ?

**Comment by pkalever** [comment]
@lxbsz This PR needs a manual rebase, please give a final push, and I will be able to merge it. Thanks!

**Comment by pkalever** [comment]
@lxbsz rebase required. Other than that, rest looks good to me. Thanks!

Tested-by: Prasanna Kumar Kalever \<prasanna.kalever@redhat.com\>
Reviewed-by: Prasanna Kumar Kalever \<prasanna.kalever@redhat.com\>

**Comment by pkalever** [comment]
@lxbsz Just a hint, --3way merge works with no conflict.

**Comment by lxbsz** [comment]
@pkalever After pulling the newest upstream code, there is no conflict. Did I miss something ?
```

[root@rhel3 gluster-block]# git log --oneline
9c1f9d7 dyn-config: read the config file line by line instead of allocating a fixed 32K buffer
b58a65b dyn-config: fix memleak for GB_LOG_LEVEL
fd62ac4 gluster-block: dump all failure msgs to stderr
9f705e7 Fix memleak for errMsg
6177f19 audit: reduce indentation for smoother flow (#133)
3ab7b06 all: fix build warnings
a651e1a spec-file: fix the error in changelog date (#139)
a79e4e6 build: move Dockerfile.buildtest to ./extras/docker/ (#143)
bb046e5 Add travis-integration for build verification using docker (#142)
6bc89f4 systemd: clear the target config only when tcmu-runner is running
f7a628f scripts: use TCMU_LOGDIR instead of GB_LOGDIR
080545c upgrade-script: clean status file on source and rpm uninstall
8ae5851 upgrade-script: Makefile and rpm spec file changes
f6b40cc extras: add a script for upgrade time activities
69f567a doc: update documentation about 'no-remote-rpc' option
64027c0 daemon: add 'no-remote-rpc' option
...

```

**Comment by pkalever** [comment]
@lxbsz looks like your rebase while adding tags fixed the Issue I saw before. Thanks the PR is now merged.

**Pull request merged**